validations/src/validations.rs

```diff
@@ -918,117 +886,152 @@ pub fn validate_tally_transaction<'a>(
+    // Slashed are liars plus non revealers
+    let liars_count = slashed_count - (commits_count - reveals_count);
+
+    let expected_tally_change = calculate_tally_change(
+        commits_count,
+        reveals_count,
+        reveals_count - liars_count,
+        &dr_state.data_request,
+    );
+
     let mut pkh_rewarded: HashSet<PublicKeyHash> = HashSet::default();
-    let witness_reward = dr_state.data_request.witness_reward;
-    let reveal_fee = dr_state.data_request.reveal_fee;
-    let commit_fee = dr_state.data_request.commit_fee;
+    let mut total_tally_value = 0;
     for (i, output) in ta_tx.outputs.iter().enumerate() {
-        if change_required && i == ta_tx.outputs.len() - 1 && output.pkh == dr_state.pkh {
-            let expected_tally_change = if commits_count == 0 {
-                witnesses as u64 * (witness_reward + reveal_fee + commit_fee)
-            } else {
-                witness_reward * dishonest_len as u64
-                    + reveal_fee * (witnesses - reveals_count) as u64
-            };
-
-            if expected_tally_change != output.value {
+        // Validation of tally change value
+        if expected_tally_change > 0 && i == ta_tx.outputs.len() - 1 && output.pkh == dr_state.pkh {
+            if output.value != expected_tally_change {
                 return Err(TransactionError::InvalidTallyChange {
                     change: output.value,
                     expected_change: expected_tally_change,
                 }
                 .into());
             }
         } else {
+            // Validation of each rewarded address was a revealer
             if dr_state.info.reveals.get(&output.pkh).is_none() {
                 return Err(TransactionError::RevealNotFound.into());
             }
-            if dishonest_pkhs.contains(&output.pkh) {
+
+            // Validation of each rewarded address was not a liar
+            if slashed_hs.contains(&output.pkh) {
                 return Err(TransactionError::DishonestReward.into());
             }
+
+            // Validation of the reward is according to the DataRequestOutput
+            if output.value != dr_state.data_request.witness_reward {
+                return Err(TransactionError::InvalidReward {
+                    value: output.value,
+                    expected_value: dr_state.data_request.witness_reward,
+                }
+                .into());
+            }
+
+            // Validation of a honest witness would not be rewarded more than once
             if pkh_rewarded.contains(&output.pkh) {
                 return Err(TransactionError::MultipleRewards { pkh: output.pkh }.into());
             }
             pkh_rewarded.insert(output.pkh);
         }
+        total_tally_value += output.value;
     }
 
-    Ok(())
+    let expected_dr_value = dr_state.data_request.checked_total_value()?;
+    let found_dr_value = dr_state.info.commits.len() as u64 * dr_state.data_request.commit_fee
```

Comment:
This could be simplified if we just sum the total value of the outputs, that value must be equal to the `expected_dr_value`.